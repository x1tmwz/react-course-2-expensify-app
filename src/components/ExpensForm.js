import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


export default class ExpensForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note:  props.expense ? props.expense.note : '',
            amount:  props.expense ? (props.expense.amount / 100).toString() : '',
            createAt: props.expense ? moment(props.expense.createAt) : moment(),
            calendarFocused: false,
            error: ''
        }

    }
    
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));

    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        const reg = /^\d{1,}(\.\d{0,2})?$/;
        if (!amount || amount.match(reg)) {
            this.setState(() => ({ amount }));
        }
    }
    onDateChange = (createAt) => {
        if (createAt) {
            this.setState(() => ({ createAt }))
        }

    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    render() {
        return (
            <div>
                {this.state.error || <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type='text'
                        onChange={this.onDescriptionChange}
                        autoFocus
                        placeholder='description'
                        value = {this.state.description}
                    />
                    <input
                        type='text'
                        onChange={this.onAmountChange}
                        placeholder='amount'
                        value = {this.state.amount} 
                        />

                    <SingleDatePicker
                        date={this.state.createAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => { return false }}
                        displayFormat={"DD/MM/YYYY"}
                    />
                    <textarea
                        onChange={this.onNoteChange}
                        value = {this.state.note}
                        placeholder='enter you notes'
                    />
                    <button type='submit'>Done!</button>
                </form>
            </div>
        )

    }
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            const message = 'please provide description and amount';
            this.setState(() => ({ error: message }));
        }
        else {
            const message = '';
            this.setState(() => ({ error: message }));
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10)*100,
                createAt:this.state.createAt.valueOf(),
                note:this.state.note
            });
        }
    }
}