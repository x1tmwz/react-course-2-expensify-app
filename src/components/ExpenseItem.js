import React from 'react';
import { Link } from 'react-router-dom';


export  const ExpenseItem = (props) => (
    <div>
        <Link to={`/edit/${props.id}`}>
            <h3>Expense : {props.description}</h3>
        </Link>
        <ul>
            {props.note && <li>note : {props.note}</li>}
            <li>amount: {props.amount}</li>
            <li>create at: {props.createAt}</li>
        </ul>
        
    </div>
)
export default ExpenseItem;