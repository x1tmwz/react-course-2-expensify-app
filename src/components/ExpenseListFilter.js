import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setFillter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilter extends React.Component {
    state = {
        calendarFocused: null,
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }
    onTextChange = (e) => {
        const text = e.target.value.trim();
        this.props.setFillter(text);
    };
    onSelectChange = (e) => {
        const sortBy = e.target.value;
        if (sortBy === 'amount') this.props.sortByAmount();
        if (sortBy === 'date') this.props.sortByDate();
    }
    render() {
        return (
            <div>
                <input
                    type='text'
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    onChange={this.onSelectChange}
                    value={this.props.filters.sortBy}
                >
                    <option value='amount'>sort by amount</option>
                    <option value='date'>sort by Date</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    displayFormat={"DD/MM/YYYY"}
                    numberOfMonths={1}
                    isOutsideRange={() => { return false }}
                    showClearDates={true}
                />
            </div>

        )
    }
}
const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};
const mapDispatchToProps = (dispatch) => ({
    setFillter: (text) => dispatch(setFillter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))

})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);