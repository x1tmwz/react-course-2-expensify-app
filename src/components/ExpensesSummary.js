import React from 'react';
import getTotalExpense from '../selectors/expense-total';
import { connect } from 'react-redux';
import numeral from 'numeral';


export const ExpensesSummary = (props) => {
    const totalAmount = getTotalExpense(props.expenses);
    return (
        <div>
            {props.expenses.length > 0 &&
                <p>Viewing {props.expenses.length} {props.expenses.length === 1 ? 'expense' : 'expenses'} totalling {numeral(totalAmount / 100).format('$0,0.00')} </p>
            }
        </div>

    );
}
const mapStateToProps = (state) => ({
    expenses: state.expenses
})
export default connect(mapStateToProps)(ExpensesSummary)



