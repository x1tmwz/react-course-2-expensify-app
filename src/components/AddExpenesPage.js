import React from 'react';
import { connect } from 'react-redux';
import ExpensForm from './ExpensForm';
import { addExpense } from '../actions/expenses';

export class AddExpenesPage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push("/");
    }
    render() {
        return (
            <div>
                <h1>New Expense</h1>
                <ExpensForm
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})


export default connect(undefined, mapDispatchToProps)(AddExpenesPage);