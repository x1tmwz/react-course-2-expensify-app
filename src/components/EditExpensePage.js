import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpnese } from '../actions/expenses';
import ExpensForm from './ExpensForm';

export class EditExpensePage extends React.Component {

    startRemoveExpnese = () => {
        this.props.startRemoveExpnese(this.props.expense.id);
        this.props.history.push("/");
    }
    startEditExpense = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpensForm
                        expense={this.props.expense}
                        onSubmit={(expense) => { this.startEditExpense(expense) }}
                    />
                    <button className="button button--secondary" onClick={this.startRemoveExpnese}>Remove Expense</button>
                </div>

            </div>
        );
    }
}
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
};
const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpnese: (id) => dispatch(startRemoveExpnese({ id }))
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);