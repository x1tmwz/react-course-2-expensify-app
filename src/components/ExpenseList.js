import React from 'react';
import ExpenseItem from './ExpenseItem';
import getVisibleExpenses from '../selectors/expenses';
import { connect } from 'react-redux';


export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (<p>No Expenses</p>) :
                (props.expenses.map((expense) => (<ExpenseItem key={expense.id} {...expense}/>)))
        }
    </div>
);
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);

