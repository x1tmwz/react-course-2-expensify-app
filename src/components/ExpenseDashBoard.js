import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashBoard = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilter />
        <ExpenseList />
    </div>
);

export default ExpenseDashBoard;