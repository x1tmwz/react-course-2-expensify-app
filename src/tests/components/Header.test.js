import { shallow } from 'enzyme';
import React from 'react';
import { Header } from '../../components/Header';

let wrapper,startLogout
beforeEach(()=>{
    startLogout = jest.fn();
    wrapper = shallow(<Header startLogout={startLogout} />);

})

test('shuold render header correctley', () => { 
    expect(wrapper).toMatchSnapshot();
});
test('should call startlogout on click',()=>{
    wrapper.find('button').simulate('click')
    expect(startLogout).toHaveBeenCalled();
})