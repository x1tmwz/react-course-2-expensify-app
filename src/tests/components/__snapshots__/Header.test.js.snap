// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shuold render header correctley 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <NavLink
    activeClassName="is-active"
    exact={true}
    to="/"
  >
    go home
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/create"
  >
    add page
  </NavLink>
  <button
    onClick={[Function]}
  >
    Logout
  </button>
</header>
`;
